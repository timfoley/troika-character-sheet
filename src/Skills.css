.skills {
    width    : 100%;
    grid-area: skills;
}

.skills input {
    margin-bottom: 0px;
    border-radius: 0;
}

.skills h2 {
    margin-top: 0px;
}

.skills__grid {
    display              : grid;
    grid-gap             : 8px;
    grid-template-columns: 1fr 36px 36px 36px;
}

.skills__grid__heading::after {
    position   : relative;
    left       : 6px;
    font-weight: normal;
}

.skills__grid__heading.skill__box--rank::after {
    content: "+";
}

.skills__grid__heading.skill__box--skill::after {
    content: "=";
}

.skill__name>input {
    border       : none;
    border-bottom: 1px solid black;
    width        : 100%;
}

.skill__name {
    grid-column: 1;
}

.skill__box {
    display        : flex;
    justify-content: center;
    align-items    : center;
    text-align     : center;
    border         : 1px solid black;
}

.skill__box--rank {
    grid-column: 2;
    width      : 100%;
    font-weight: bold;
}

.skill__box--skill-stat {
    grid-column: 3;
}

.skill__box--total {
    grid-column: 4;
    font-weight: bold;
    color      : tomato;
    font-size  : 1.5rem;
}

.skills__grid__heading {
    font-weight: normal;
    color      : black;
    text-align : right;
    font-size  : .8rem;
}

.skills .header {
    font-size: .5rem;
}