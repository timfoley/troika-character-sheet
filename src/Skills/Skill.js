import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";

export const Skill = props => {
	const skillNameInput = useRef(null);

	useEffect(() => {
		props.shouldFocus && skillNameInput.current.focus();
	}, [props.shouldFocus]);

	const setName = e => props.setSkillName(props.index, e.target.value);
	const setRank = e =>
		props.setSkillRank(props.index, parseInt(e.target.value));

	const getTotalClass = rank => {
		let className = "skill__box skill__box--total ";
		return rank === 0 ? className + "zero-to-hide" : className;
	};

	return (
		<React.Fragment>
			<div className="skill__name">
				<input
					ref={skillNameInput}
					aria-label="skill name"
					className="skill__name__input"
					type="text"
					value={props.skill.name}
					onChange={setName}
				/>
			</div>
			<input
				aria-label="skill rank"
				className={`skill__box skill__box--rank hide-spinners ${
					props.skill.rank === 0 ? "zero-to-hide" : ""
				}`}
				type="number"
				value={props.skill.rank}
				onChange={setRank}
			/>
			<div
				aria-label="skill stat"
				className={`skill__box skill__box--skill-stat ${
					props.skillStat === 0 ? "zero-to-hide" : ""
				}`}
			>
				{isNaN(props.skillStat) ? "-" : props.skillStat}
			</div>
			<div
				aria-label="skill total"
				className={getTotalClass(props.skill.rank)}
			>
				{isNaN(props.skillStat)
					? "-"
					: props.skillStat + props.skill.rank}
			</div>
		</React.Fragment>
	);
};

Skill.propTypes = {
	skill: PropTypes.object.isRequired,
	index: PropTypes.number.isRequired,
	setSkillName: PropTypes.func.isRequired,
	setSkillRank: PropTypes.func.isRequired,
	skillStat: PropTypes.number,
};
