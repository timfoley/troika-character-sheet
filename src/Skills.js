import React from "react";
import PropTypes from "prop-types";

import "./Skills.css";

export function Skills(props) {
	const setSkillName = (i, name) => {
		const newSkills = props.skills;
		newSkills[i] = { ...newSkills[i], name };
		props.updateCharacter("skills", newSkills);
	};

	const setSkillRank = (i, rank) => {
		const newSkills = props.skills;
		newSkills[i] = {
			...newSkills[i],
			rank,
		};
		props.updateCharacter("skills", newSkills);
	};

	const addSkill = () => {
		props.updateCharacter("skills", [
			...props.skills,
			{ name: "", rank: 0 },
		]);
	};

	const handleAddSkillKeyPress = e => {
		if (e.which === 32 || e.key === "enter") {
			addSkill();
		}
	};

	const getTotalClass = rank => {
		let className = "skill__box skill__box--total ";
		return rank === 0 ? className + "zero-to-hide" : className;
	};

	return (
		<div className="skills">
			<h2>Advanced Skills and Spells</h2>
			<div className="skills__grid">
				<div className="skills__grid__heading skill__box--rank">
					Rank
				</div>
				<div className="skills__grid__heading skill__box--skill">
					Skill
				</div>
				<div className="skills__grid__heading skill__box--total">
					Total
				</div>
				{props.skills.map((skill, index) => (
					<React.Fragment key={skill + index}>
						<div className="skill__name">
							<input
								aria-label="skill name"
								className="skill__name__input"
								type="text"
								value={skill.name}
								onChange={e =>
									setSkillName(index, e.target.value)
								}
							/>
						</div>
						<input
							aria-label="skill rank"
							className={`skill__box skill__box--rank hide-spinners ${
								skill.rank === 0 ? "zero-to-hide" : ""
							}`}
							type="number"
							value={skill.rank}
							onChange={e =>
								setSkillRank(index, parseInt(e.target.value))
							}
						/>
						<div
							aria-label="skill stat"
							className={`skill__box skill__box--skill-stat ${
								props.skillStat === 0 ? "zero-to-hide" : ""
							}`}
						>
							{isNaN(props.skillStat) ? "-" : props.skillStat}
						</div>
						<div
							aria-label="skill total"
							className={getTotalClass(skill.rank)}
						>
							{isNaN(props.skillStat)
								? "-"
								: props.skillStat + skill.rank}
						</div>
					</React.Fragment>
				))}
			</div>
			<div
				className="skills__add-button clickable"
				area-label="click to add a skill"
				onClick={addSkill}
				onKeyPress={handleAddSkillKeyPress}
				tabIndex="0"
			>
				Add Skill
			</div>
		</div>
	);
}

Skills.propTypes = {
	skills: PropTypes.array,
	skillStat: PropTypes.number,
	updateCharacter: PropTypes.func,
};

Skills.defaultProps = {
	skills: [
		{ name: "", rank: 0 },
		{ name: "", rank: 0 },
		{ name: "", rank: 0 },
		{ name: "", rank: 0 },
		{ name: "", rank: 0 },
	],
	skillStat: 0,
};
