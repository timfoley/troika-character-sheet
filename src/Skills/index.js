import React, { useState } from "react";
import PropTypes from "prop-types";

import "./Skills.css";

import { Skill } from "./Skill";

export function Skills(props) {
	const [indexToFocus, setIndexToFocus] = useState(null);

	const setSkillName = (i, name) => {
		const newSkills = props.skills;
		newSkills[i] = { ...newSkills[i], name };
		props.updateCharacter("skills", newSkills);
	};

	const setSkillRank = (i, rank) => {
		const newSkills = props.skills;
		newSkills[i] = {
			...newSkills[i],
			rank,
		};
		props.updateCharacter("skills", newSkills);
	};

	const addSkill = () => {
		setIndexToFocus(props.skills.length);
		props.updateCharacter("skills", [
			...props.skills,
			{ name: "", rank: 0 },
		]);
	};

	const handleAddSkillKeyPress = e => {
		if (e.which === 32 || e.key === "enter") {
			addSkill();
		}
	};

	return (
		<div className="skills">
			<h2>Advanced Skills and Spells</h2>
			<div className="skills__grid">
				<div className="skills__grid__heading skill__box--rank">
					Rank
				</div>
				<div className="skills__grid__heading skill__box--skill">
					Skill
				</div>
				<div className="skills__grid__heading skill__box--total">
					Total
				</div>
				{props.skills.map((skill, index) => (
					<Skill
						key={skill + index}
						skill={skill}
						index={index}
						setSkillName={setSkillName}
						setSkillRank={setSkillRank}
						skillStat={props.skillStat}
						shouldFocus={indexToFocus === index}
					/>
				))}
			</div>
			<div
				className="skills__add-button clickable"
				area-label="click to add a skill"
				onClick={addSkill}
				onKeyPress={handleAddSkillKeyPress}
				tabIndex="0"
			>
				Add Skill
			</div>
		</div>
	);
}

Skills.propTypes = {
	skills: PropTypes.array,
	skillStat: PropTypes.number,
	updateCharacter: PropTypes.func,
};

Skills.defaultProps = {
	skills: [
		{ name: "", rank: 0 },
		{ name: "", rank: 0 },
		{ name: "", rank: 0 },
		{ name: "", rank: 0 },
		{ name: "", rank: 0 },
	],
	skillStat: 0,
};
