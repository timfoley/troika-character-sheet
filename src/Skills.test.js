import React from "react";
import ReactDOM from "react-dom";
import { render } from "@testing-library/react";

import { Skills } from "./Skills";

it("renders without crashing", () => {
	const div = document.createElement("div");
	ReactDOM.render(<Skills />, div);
	ReactDOM.unmountComponentAtNode(div);
});

it("matches snapshot", () => {
	const { container } = render(<Skills />);
	expect(container.firstChild).toMatchSnapshot();
});
