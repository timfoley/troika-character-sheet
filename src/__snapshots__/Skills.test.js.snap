// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  class="skills"
>
  <h2>
    Advanced Skills and Spells
  </h2>
  <div
    class="skills__grid"
  >
    <div
      class="skills__grid__heading skill__box--rank"
    >
      Rank
    </div>
    <div
      class="skills__grid__heading skill__box--skill"
    >
      Skill
    </div>
    <div
      class="skills__grid__heading skill__box--total"
    >
      Total
    </div>
    <div
      class="skill__name"
    >
      <input
        aria-label="skill name"
        class="skill__name__input"
        type="text"
        value=""
      />
    </div>
    <input
      aria-label="skill rank"
      class="skill__box skill__box--rank hide-spinners"
      type="number"
      value="1"
    />
    <div
      aria-label="skill stat"
      class="skill__box skill__box--skill-stat"
    >
      0
    </div>
    <div
      aria-label="skill total"
      class="skill__box skill__box--total"
    >
      1
    </div>
    <div
      class="skill__name"
    >
      <input
        aria-label="skill name"
        class="skill__name__input"
        type="text"
        value=""
      />
    </div>
    <input
      aria-label="skill rank"
      class="skill__box skill__box--rank hide-spinners"
      type="number"
      value="1"
    />
    <div
      aria-label="skill stat"
      class="skill__box skill__box--skill-stat"
    >
      0
    </div>
    <div
      aria-label="skill total"
      class="skill__box skill__box--total"
    >
      1
    </div>
    <div
      class="skill__name"
    >
      <input
        aria-label="skill name"
        class="skill__name__input"
        type="text"
        value=""
      />
    </div>
    <input
      aria-label="skill rank"
      class="skill__box skill__box--rank hide-spinners"
      type="number"
      value="1"
    />
    <div
      aria-label="skill stat"
      class="skill__box skill__box--skill-stat"
    >
      0
    </div>
    <div
      aria-label="skill total"
      class="skill__box skill__box--total"
    >
      1
    </div>
    <div
      class="skill__name"
    >
      <input
        aria-label="skill name"
        class="skill__name__input"
        type="text"
        value=""
      />
    </div>
    <input
      aria-label="skill rank"
      class="skill__box skill__box--rank hide-spinners"
      type="number"
      value="1"
    />
    <div
      aria-label="skill stat"
      class="skill__box skill__box--skill-stat"
    >
      0
    </div>
    <div
      aria-label="skill total"
      class="skill__box skill__box--total"
    >
      1
    </div>
    <div
      class="skill__name"
    >
      <input
        aria-label="skill name"
        class="skill__name__input"
        type="text"
        value=""
      />
    </div>
    <input
      aria-label="skill rank"
      class="skill__box skill__box--rank hide-spinners"
      type="number"
      value="1"
    />
    <div
      aria-label="skill stat"
      class="skill__box skill__box--skill-stat"
    >
      0
    </div>
    <div
      aria-label="skill total"
      class="skill__box skill__box--total"
    >
      1
    </div>
  </div>
  <div
    area-label="click to add a skill"
    class="skills__add-button clickable"
    tabindex="0"
  >
    Add Skill
  </div>
</div>
`;
